### PR TITLE
Add more model to the domif-setlink test

### PR DESCRIPTION
**Pull request by yalzhang**
The set link up and down are related with the interface driver.
So expand the test coverage to test all current supported models.

Signed-off-by: Yalan <yalzhang@redhat.com>

### CODE REVIEW

**Comment by yalzhang** [comment]
run all the cases locally, the result is as expected:
$ avocado run --vt-type libvirt virsh.domif_setlink_getlink
JOB ID     : 2777aed3ae6f22fe69d357c02801717b1719e34f
JOB LOG    : /root/avocado/job-results/job-2021-08-04T20.05-2777aed/job.log
 (01/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_net.no_config.domain_name.running_guest.domif_setlink.setlink_up: PASS (62.02 s)
 (02/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_net.no_config.domain_name.running_guest.domif_setlink.setlink_down: PASS (64.11 s)
 (03/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_net.no_config.domain_ID.running_guest.domif_setlink.setlink_up: PASS (61.60 s)
 (04/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_mac.no_config.domain_name.running_guest.domif_setlink.setlink_up: PASS (60.97 s)
 (05/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_mac.no_config.domain_name.running_guest.domif_setlink.setlink_down: PASS (64.55 s)
 (06/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_mac.no_config.domain_name.shutoff_guest.domif_setlink.setlink_up: PASS (34.94 s)
 (07/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_mac.no_config.domain_name.shutoff_guest.domif_setlink.setlink_down: PASS (38.10 s)
 (08/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_mac.no_config.domain_UUID.shutoff_guest.domif_setlink.setlink_up: PASS (34.96 s)
 (09/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_mac.with_config.domain_name.running_guest.domif_setlink.setlink_up: PASS (85.92 s)
 (10/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_mac.with_config.domain_name.running_guest.domif_setlink.setlink_down: PASS (88.85 s)
 (11/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_mac.with_persistent.domain_name.running_guest.domif_setlink.setlink_up: PASS (86.51 s)
 (12/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.virtio.interface_mac.with_persistent.domain_name.running_guest.domif_setlink.setlink_down: PASS (88.62 s)
 (13/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_net.no_config.domain_name.running_guest.domif_setlink.setlink_up: PASS (65.56 s)
 (14/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_net.no_config.domain_name.running_guest.domif_setlink.setlink_down: PASS (68.60 s)
 (15/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_net.no_config.domain_ID.running_guest.domif_setlink.setlink_up: PASS (65.08 s)
 (16/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_mac.no_config.domain_name.running_guest.domif_setlink.setlink_up: PASS (64.79 s)
 (17/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_mac.no_config.domain_name.running_guest.domif_setlink.setlink_down: PASS (71.05 s)
 (18/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_mac.no_config.domain_name.shutoff_guest.domif_setlink.setlink_up: PASS (39.82 s)
 (19/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_mac.no_config.domain_name.shutoff_guest.domif_setlink.setlink_down: FAIL: Link state shouldn't be down in guest (37.06 s)
 (20/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_mac.no_config.domain_UUID.shutoff_guest.domif_setlink.setlink_up: PASS (40.00 s)
 (21/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_mac.with_config.domain_name.running_guest.domif_setlink.setlink_up: PASS (93.90 s)
 (22/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_mac.with_config.domain_name.running_guest.domif_setlink.setlink_down: FAIL: Link state should be down in guest (89.27 s)
 (23/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_mac.with_persistent.domain_name.running_guest.domif_setlink.setlink_up: PASS (93.70 s)
 (24/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.e1000e.interface_mac.with_persistent.domain_name.running_guest.domif_setlink.setlink_down: FAIL: Link state shouldn't be down in guest (89.79 s)
 (25/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_net.no_config.domain_name.running_guest.domif_setlink.setlink_up: PASS (60.31 s)
 (26/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_net.no_config.domain_name.running_guest.domif_setlink.setlink_down: PASS (64.80 s)
 (27/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_net.no_config.domain_ID.running_guest.domif_setlink.setlink_up: PASS (61.25 s)
 (28/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_mac.no_config.domain_name.running_guest.domif_setlink.setlink_up: PASS (63.15 s)
 (29/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_mac.no_config.domain_name.running_guest.domif_setlink.setlink_down: PASS (65.59 s)
 (30/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_mac.no_config.domain_name.shutoff_guest.domif_setlink.setlink_up: PASS (36.52 s)
 (31/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_mac.no_config.domain_name.shutoff_guest.domif_setlink.setlink_down: PASS (38.53 s)
 (32/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_mac.no_config.domain_UUID.shutoff_guest.domif_setlink.setlink_up: PASS (36.66 s)
 (33/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_mac.with_config.domain_name.running_guest.domif_setlink.setlink_up: PASS (88.02 s)
 (34/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_mac.with_config.domain_name.running_guest.domif_setlink.setlink_down: PASS (91.39 s)
 (35/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_mac.with_persistent.domain_name.running_guest.domif_setlink.setlink_up: PASS (89.68 s)
 (36/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.positive_test.rtl8139.interface_mac.with_persistent.domain_name.running_guest.domif_setlink.setlink_down: PASS (91.01 s)
 (37/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.multi_functional_test.virtio.check_update_device: PASS (73.29 s)
 (38/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.multi_functional_test.virtio.check_options_config: PASS (85.31 s)
 (39/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.multi_functional_test.e1000e.check_update_device: FAIL: Check update_device failed (69.25 s)
 (40/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.multi_functional_test.e1000e.check_options_config: PASS (94.14 s)
 (41/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.multi_functional_test.rtl8139.check_update_device: PASS (73.28 s)
 (42/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.multi_functional_test.rtl8139.check_options_config: PASS (88.63 s)
 (43/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.running_guest_invalid_option: PASS (60.77 s)
 (44/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.shutoff_guest_with_vnet: PASS (34.68 s)
 (45/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.config_option_with_vnet: PASS (86.53 s)
 (46/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.no_exist_device.with_config_option.no_exist_interface_net: PASS (84.77 s)
 (47/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.no_exist_device.with_config_option.no_exist_interface_mac: PASS (84.67 s)
 (48/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.no_exist_device.without_config_option.no_exist_interface_net: PASS (59.58 s)
 (49/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.no_exist_device.without_config_option.no_exist_interface_mac: PASS (59.04 s)
 (50/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.error_operation: PASS (58.92 s)
 (51/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.no_operation: PASS (59.33 s)
 (52/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.no_device: PASS (59.51 s)
 (53/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.no_match_domain_uuid: PASS (59.35 s)
 (54/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.no_match_domain_name: PASS (60.91 s)
 (55/55) type_specific.io-github-autotest-libvirt.virsh.domif_setlink_getlink.negative_test.without_domain_param: PASS (60.48 s)
RESULTS    : PASS 51 | ERROR 0 | FAIL 4 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 3704.86 s

The 4 failed case is caused by an existing bug 1707441. I will add them into the blacklist.

File: libvirt/tests/src/virsh_cmd/domain/virsh_domif_setlink_getlink.py [hunk 1]
```diff
@@ -222,29 +230,36 @@ def check_update_device(vm, if_name, session):
...
                         guest_if_state(guest_if_name, session)):
                     error_msg = "Link state should be down in guest"
+                if error_msg:
+                    test.fail(error_msg)
 
             # Test of setting link state by update_device command
             if check_link_by_update_device:
                 if not check_update_device(vm, guest_if_name, session):
                     error_msg = "Check update_device failed"
+                    if error_msg:
```
Comment (Yingshun) [inline libvirt/tests/src/virsh_cmd/domain/virsh_domif_setlink_getlink.py:240]: I think you could raise testFail directly. what do you think?

File: libvirt/tests/src/virsh_cmd/domain/virsh_domif_setlink_getlink.py [hunk 2]
```diff
@@ -222,29 +230,36 @@ def check_update_device(vm, if_name, session):
...
+                        test.fail(error_msg)
 
             # Set the link up make host connect with vm
             domif_setlink(vm_name[0], device, "up", "")
             if not utils_misc.wait_for(
                     lambda: guest_if_state(guest_if_name, session), 5):
                 error_msg = "Link state isn't up in guest"
+                if error_msg:
+                    test.fail(error_msg)
```
Comment (Yingshun) [inline libvirt/tests/src/virsh_cmd/domain/virsh_domif_setlink_getlink.py:249]: same as above.

**Review by Yingshun — COMMENTED** [review]
Others are good to me.

File: libvirt/tests/cfg/virsh_cmd/domain/virsh_domif_setlink_getlink.cfg [hunk 3]
```diff
@@ -44,6 +45,13 @@
                     if_name = "vnet0"
                 - interface_mac:
                     if_device = "mac"
+            variants:
+                - virtio:
+                    model_type = "virtio"
+                - e1000e:
+                    model_type = "e1000e"
+                - rtl8139:
+                    model_type = "rtl8139"
```
[the remark this answers is not in the document]
Comment (yalzhang) [reply libvirt/tests/cfg/virsh_cmd/domain/virsh_domif_setlink_getlink.cfg:54]: Good question, I will check it.

**Review by chunfuwen — APPROVED** [review]
LGTM

File: libvirt/tests/cfg/virsh_cmd/domain/virsh_domif_setlink_getlink.cfg (hunk 3 above)
[the remark this answers is not in the document]
Comment (yalzhang) [reply libvirt/tests/cfg/virsh_cmd/domain/virsh_domif_setlink_getlink.cfg:54]: I have checked and found for other archs, they only support virtio model type.

File: libvirt/tests/src/virsh_cmd/domain/virsh_domif_setlink_getlink.py (hunk 1 above)
Comment (yalzhang) [reply libvirt/tests/src/virsh_cmd/domain/virsh_domif_setlink_getlink.py]: Updated, Thank you!

File: libvirt/tests/src/virsh_cmd/domain/virsh_domif_setlink_getlink.py (hunk 2 above)
Comment (yalzhang) [reply libvirt/tests/src/virsh_cmd/domain/virsh_domif_setlink_getlink.py]: Updated

File: libvirt/tests/cfg/virsh_cmd/domain/virsh_domif_setlink_getlink.cfg (hunk 3 above)
[the remark this answers is not in the document]
Comment (yalzhang) [reply libvirt/tests/cfg/virsh_cmd/domain/virsh_domif_setlink_getlink.cfg]: Have add "only x86_64" to the cfg file under the e1000e and rtl8139. Hope it works. :-)

**Review by Yingshun — APPROVED** [review]

File: libvirt/tests/cfg/virsh_cmd/domain/virsh_domif_setlink_getlink.cfg [hunk 4]
```diff
@@ -58,7 +68,15 @@
                     if_operation = "up"
                     if_opitons = "--config"
                     if_device = "mac"
+            variants:
+                - virtio:
+                    model_type = "virtio"
+                - e1000e:
+                    model_type = "e1000e"
+                - rtl8139:
+                    model_type = "rtl8139"
```
Comment (chloerh) [inline libvirt/tests/cfg/virsh_cmd/domain/virsh_domif_setlink_getlink.cfg:77]: only x86 too?

Comment (yalzhang) [reply libvirt/tests/cfg/virsh_cmd/domain/virsh_domif_setlink_getlink.cfg]: Updated, Thank you! I can't believe that I made such rookie mistake again. :-) hahaha~~

**Review by chloerh — APPROVED** [review]

**Pull request merged**